### PR TITLE
Closure command handler

### PR DESCRIPTION
**Pull request by francescotrucchia**
This command handler simplify the mapping of command on class. Instead of using method mapping it use closure. In this way it's also possibile to add new command handler runtime.

### CODE REVIEW

File: src/Broadway/CommandHandling/ClosureCommandHandler.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        $reflection = new \ReflectionFunction($handler);
+        $reflectionParams = $reflection->getParameters();
+
+        if(!isset($reflectionParams[0]) || !$reflectionParams[0]->getClass()) {
+            throw new CommandNotAnObjectException('');
+        }
+
+        $classParts = explode('\\', $reflectionParams[0]->getClass()->name);
+        $index = end($classParts);
```
Comment (othillo) [inline src/Broadway/CommandHandling/ClosureCommandHandler.php]: $index = $reflectionParams[0]->getClass()->getShortName();

File: src/Broadway/CommandHandling/ClosureCommandHandler.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
+<?php
+
+
```
Comment (othillo) [inline src/Broadway/CommandHandling/ClosureCommandHandler.php]: can you add our license block?

File: src/Broadway/CommandHandling/ClosureCommandHandler.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+namespace Broadway\CommandHandling;
+
+
```
Comment (othillo) [inline src/Broadway/CommandHandling/ClosureCommandHandler.php]: remove line

File: src/Broadway/CommandHandling/ClosureCommandHandler.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        $classParts = explode('\\', get_class($command));
+        $index = end($classParts);
+
+        if(!isset($this->handlers[$index])) {
+            return;
+        }
+
+        $this->handlers[$index]($command);
+    }
+}
```
Comment (othillo) [inline src/Broadway/CommandHandling/ClosureCommandHandler.php]: newline

File: src/Broadway/CommandHandling/ClosureCommandHandler.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    public function handle($command)
+    {
+        if (! is_object($command)) {
+            throw new CommandNotAnObjectException();
+        }
+
+        $classParts = explode('\\', get_class($command));
+        $index = end($classParts);
```
Comment (othillo) [inline src/Broadway/CommandHandling/ClosureCommandHandler.php]: you may want to use ReflectionClass->getShortName here, just a matter of taste

File: src/Broadway/CommandHandling/ClosureCommandHandler.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    private $handlers;
+
+    public function __construct()
+    {
+        $this->configure();
+    }
+
+    abstract protected function configure();
+
+    public function add(\Closure $handler)
```
Comment (othillo) [inline src/Broadway/CommandHandling/ClosureCommandHandler.php]: add a docblock, please

File: test/Broadway/CommandHandling/ClosureCommandHandlerTest.php [hunk 7]
```diff
@@ -0,0 +1,83 @@
+<?php
+
+
```
Comment (othillo) [inline test/Broadway/CommandHandling/ClosureCommandHandlerTest.php]: add license block

File: test/Broadway/CommandHandling/ClosureCommandHandlerTest.php [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        $this->assertTrue($commandHandler->handled);
+    }
+
+    /**
+     * @test
+     *
+     * @dataProvider unresolvableCommands
+     */
+    public function handle_should_throw_exception_when_impossible_to_delegate_to_a_valid_closure($command)
```
Comment (othillo) [inline test/Broadway/CommandHandling/ClosureCommandHandlerTest.php]: what you're testing here is that the handle method only accepts object, right?

suggestion: it_throws_when_handling_a_non_object_command()

File: test/Broadway/CommandHandling/ClosureCommandHandlerTest.php [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        $command        = new ClosureCommandHandlerTestCommand();
+        $commandHandler->handle($command);
+
+        $this->assertTrue($commandHandler->handled);
+    }
+
+    /**
+     * @test
+     *
+     * @dataProvider unresolvableCommands
```
Comment (othillo) [inline test/Broadway/CommandHandling/ClosureCommandHandlerTest.php]: the data provider is a bit too much to my taste, using one test case with a non object demonstrates the point, IMO.

File: src/Broadway/CommandHandling/ClosureCommandHandler.php [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        $reflection = new \ReflectionFunction($handler);
+        $reflectionParams = $reflection->getParameters();
+
+        if(!isset($reflectionParams[0]) || !$reflectionParams[0]->getClass()) {
+            throw new CommandNotAnObjectException('');
```
Comment (othillo) [inline src/Broadway/CommandHandling/ClosureCommandHandler.php]: should this be a ClosureParameterNotAnObjectException or somthing like that?

File: src/Broadway/CommandHandling/ClosureCommandHandler.php [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        $classParts = explode('\\', $reflectionParams[0]->getClass()->name);
+        $index = end($classParts);
+
+        $this->handlers[$index] = $handler;
+    }
+
+    public function handle($command)
```
Comment (othillo) [inline src/Broadway/CommandHandling/ClosureCommandHandler.php]: add a docblock, please

File: test/Broadway/CommandHandling/ClosureCommandHandlerTest.php [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    /**
+     * @test
+     */
+    public function it_delegates_command_to_proper_handle_function()
+    {
+        $commandHandler = new TestClosureCommandHandler();
+        $command        = new ClosureCommandHandlerTestCommand();
+        $commandHandler->handle($command);
+
+        $this->assertTrue($commandHandler->handled);
```
Comment (othillo) [inline test/Broadway/CommandHandling/ClosureCommandHandlerTest.php]: the handled property does not exist in the ClosureCommandHandler (or the CommandHandlerInterface)

**Comment by francescotrucchia** [comment]
@othillo i try to fix all your suggests ;-)!! Thanks a lot.

**Comment by othillo** [comment]
@francescotrucchia can you rebase on master? now there are 55 commits of which only about 6 are related to the ClosureCommandHandler

**Comment by francescotrucchia** [comment]
@othillo thanks. I did it ;-) !!

File: src/Broadway/CommandHandling/ClosureCommandHandler.php [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    /**
+     * @param mixed $command
+     */
+    public function handle($command)
+    {
+        if (! is_object($command)) {
+            throw new ClosureParameterNotAnObjectException();
+        }
+
+        $reflectionClass = new \ReflectionClass($command);
```
Comment (othillo) [inline src/Broadway/CommandHandling/ClosureCommandHandler.php]: reflection is expensive, so just wondering if can you use $index = get_class($command) here and $index = $reflectionParams[0]->getClass()->getName() in add()?

File: src/Broadway/CommandHandling/ClosureCommandHandler.php [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     * @param \Closure $handler
+     */
+    public function add(\Closure $handler)
+    {
+        $reflection = new \ReflectionFunction($handler);
+        $reflectionParams = $reflection->getParameters();
+
+        if(!isset($reflectionParams[0]) || !$reflectionParams[0]->getClass()) {
```
Comment (francescotrucchia) [inline src/Broadway/CommandHandling/ClosureCommandHandler.php]: $reflectionParams variable it's needed in the line below. I don't like to create an assert method that return a variable.

File: src/Broadway/CommandHandling/ClosureCommandHandler.php [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+        $this->handlers[$index] = $handler;
+    }
+
+    /**
+     * @param mixed $command
+     */
+    public function handle($command)
+    {
+        if (!is_object($command)) {
+            throw new ClosureParameterNotAnObjectException();
```
Comment (othillo) [inline src/Broadway/CommandHandling/ClosureCommandHandler.php]: this should be CommandNotAnObjectException

File: test/Broadway/CommandHandling/ClosureCommandHandlerTest.php [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        $this->assertTrue($command->handle);
+    }
+
+    /**
+     * @test
+     */
+    public function it_throws_when_handling_a_non_object_command()
+    {
+        $commandHandler = new ClosureCommandHandler();
+        $this->setExpectedException(ClosureParameterNotAnObjectException::class);
```
Comment (othillo) [inline test/Broadway/CommandHandling/ClosureCommandHandlerTest.php]: this should be CommandNotAnObjectException

File: test/Broadway/CommandHandling/ClosureCommandHandlerTest.php [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+        $commandHandler->handle('foo');
+    }
+}
+
```
Comment (othillo) [inline test/Broadway/CommandHandling/ClosureCommandHandlerTest.php]: can you add a test for ClosureParameterNotAnObjectException when adding a closure without an object parameter?

it_throws_when_adding_a_closure_without_an_object_argument or something like that?

File: src/Broadway/CommandHandling/ClosureCommandHandler.php [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * Using this class command can be registered with closure
+ */
+class ClosureCommandHandler implements CommandHandler
+{
+    /**
+     * @var \Closure[]
+     */
+    private $handlers;
```
Comment (othillo) [inline src/Broadway/CommandHandling/ClosureCommandHandler.php]: private $handlers = [];

File: src/Broadway/CommandHandling/Exception/ClosureParameterNotAnObjectException.php [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Broadway\CommandHandling\Exception;
+
+use InvalidArgumentException;
+
+/**
+ * Commands should be objects.
```
Comment (othillo) [inline src/Broadway/CommandHandling/Exception/ClosureParameterNotAnObjectException.php]: Closure parameters should be objects.

File: src/Broadway/CommandHandling/ClosureCommandHandler.php [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Broadway\CommandHandling;
+
+use Broadway\CommandHandling\Exception\ClosureParameterNotAnObjectException;
+use Broadway\CommandHandling\Exception\CommandNotAnObjectException;
+
+/**
+ * Using this class command can be registered with closure
```
Comment (othillo) [inline src/Broadway/CommandHandling/ClosureCommandHandler.php]: Using this class command handlers can be registered with closures?

**Comment by othillo** [comment]
just 2 comments on the docblock left. Otherwise it looks good to me!
Can you squash your commits?

**Comment by francescotrucchia** [comment]
@othillo I rebase my commits. Thanks

**Comment by francescotrucchia** [comment]
@wjzijderveld any news? Thanks

**Comment by wjzijderveld** [comment]
Thanks @francescotrucchia :-)

**Pull request merged**